staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,619 @@
+			timeoutIdx = i
+
+		} else {
+			break
+		}
+	}
+	// remove timed out packets from queue
+	if timeoutIdx != -1 {
+		// timeoutIdx + 1 to remove the last timeout pkt
+		removeIdx := timeoutIdx + 1
+
+		// remove all the timeout packets
+		queue.Requests = pkts[removeIdx:]
+		// decrement the # of requestsEnqueued
+		qs.numRequestsEnqueued -= removeIdx
+	}
+}
+
+// GetRequestsExecuting gets the # of requests which are "executing":
+// this is the# of requests/packets which have been dequeued but have not had
+// finished (via the FinishRequest method invoked after service)
+func (qs *queueSetImpl) GetRequestsExecuting() int {
+	total := 0
+	for _, queue := range qs.queues {
+		total += queue.RequestsExecuting
+	}
+	return total
+}
+
+// ChooseQueueIdx uses shuffle sharding to select a queue index
+// using a 'hashValue'.  The 'hashValue' derives a hand from a set range of
+// indexes (range 'desiredNumQueues') and returns the queue with the least queued packets
+// from a dealt hand (of size 'handSize')
+func (qs *queueSetImpl) ChooseQueueIdx(hashValue uint64) int {
+	// TODO(aaron-prindle) currently a lock is held for this in a larger anonymous function
+	// verify that makes sense...
+
+	bestQueueIdx := -1
+	bestQueueLen := int(math.MaxInt32)
+	// desiredNumQueues is used here instead of numQueues to omit quiesce queues
+	qs.dealer.Deal(hashValue, func(queueIdx int) {
+		thisLen := len(qs.queues[queueIdx].Requests)
+		if thisLen < bestQueueLen {
+			bestQueueIdx, bestQueueLen = queueIdx, thisLen
+		}
+	})
+	return bestQueueIdx
+}
+
+// rejectOrEnqueue rejects or enqueues the newly arrived request if
+// resource criteria isn't met
+func (qs *queueSetImpl) rejectOrEnqueue(packet *fq.Request) bool {
+	queue := packet.Queue
+	curQueueLength := len(queue.Requests)
+	// rejects the newly arrived request if resource criteria not met
+	if qs.GetRequestsExecuting() >= qs.concurrencyLimit &&
+		curQueueLength >= qs.queueLengthLimit {
+		return false
+	}
+
+	qs.enqueue(packet)
+	return true
+}
+
+// enqueues a packet into an queueSetImpl
+func (qs *queueSetImpl) enqueue(packet *fq.Request) {
+
+	queue := packet.Queue
+	queue.Enqueue(packet)
+	qs.updateQueueVirStartTime(packet, queue)
+	qs.numRequestsEnqueued++
+
+	metrics.UpdateFlowControlRequestsInQueue(qs.name, qs.numRequestsEnqueued)
+}
+
+// synctime is used to sync the time of the queueSetImpl by looking at the elapsed
+// time since the last sync and this value based on the 'virtualtime ratio'
+// which scales inversely to the # of active flows
+func (qs *queueSetImpl) synctime() {
+	realNow := qs.clk.Now()
+	timesincelast := realNow.Sub(qs.lastRealTime).Seconds()
+	qs.lastRealTime = realNow
+	qs.vt += timesincelast * qs.getvirtualtimeratio()
+}
+
+func (qs *queueSetImpl) getvirtualtimeratio() float64 {
+	NEQ := 0
+	reqs := 0
+	for _, queue := range qs.queues {
+		reqs += queue.RequestsExecuting
+		// It might be best to delete this line. If everything is working
```

Comment:
This comment, and the commented out statement `reqs += len(queue.Requests)`, are debris from earlier discussions.  I doubt we should retain them.

Reply:
Removed